general/mem_manager.hpp

```diff
@@ -67,119 +496,268 @@ class MemoryManager
    /// Return true if a global memory manager instance exists
    static bool Exists() { return exists; }
 
-   /** @brief Translates ptr to host or device address, depending on what
-       backends are currently allowed by the Device class and on the ptr
-       state. */
-   void *Ptr(void *ptr);
-   const void *Ptr(const void *ptr);
-
-   /// Data will be pushed/pulled before the copy happens on the H or the D
-   void* Memcpy(void *dst, const void *src,
-                std::size_t bytes, const bool async = false);
-
    /// Return the bytes of the memory region which base address is ptr
    std::size_t Bytes(const void *ptr);
 
-   /// Return true if the registered pointer is on the host side
-   bool IsOnHost(const void *ptr);
-
    /// Return true if the pointer has been registered
    bool IsKnown(const void *ptr);
 
-   /// Return true if the pointer is an alias inside a registered memory region
-   bool IsAlias(const void *ptr);
+   /// Check if pointer has been registered in the memory manager
+   void RegisterCheck(void *ptr);
 
-   /// Push the data to the device
-   void Push(const void *ptr, const std::size_t bytes =0);
+   /// Prints all pointers known by the memory manager
+   void PrintPtrs(void);
+};
 
-   /// Pull the data from the device
-   void Pull(const void *ptr, const std::size_t bytes =0);
 
-   /// Return the corresponding device pointer of ptr, allocating and moving the
-   /// data if needed (used in OccaPtr)
-   void *GetDevicePtr(const void *ptr);
+// Inline methods
 
-   /// Registers external host pointer in the memory manager which will manage
-   /// the corresponding device pointer, but not the provided host pointer.
-   template<class T>
-   void RegisterHostPtr(T *ptr_host, const std::size_t size)
+template <typename T>
+inline void Memory<T>::New(int size, MemoryType mt)
+{
+   if (mt == MemoryType::HOST)
    {
-      Insert(ptr_host, size*sizeof(T));
-#ifdef MFEM_DEBUG
-      RegisterCheck(ptr_host);
-#endif
+      New(size);
    }
+   else
+   {
+      // Allocate the host pointer with new T[] if 'mt' is a pure device memory
+      // type, e.g. CUDA.
+      T *tmp = (mt == MemoryType::CUDA) ? new T[size] : NULL;
+      h_ptr = (T*)MemoryManager::New_(tmp, size*sizeof(T), mt, flags);
```

Comment:
I don't understand everything here, but is the MemoryManager class still needed? Maybe everything could be handled by Memory, perhaps even without the hashing?

Reply:
With this PR, the `MemoryManager` is only used directly by the `Memory` class, so any changes we want to make in the `MemoryManager` and the implementation of `Memory` should not affect the rest of the code.